### ISSUE TITLE
Verify this code works on an SBC (ie. Raspberry Pi)

### ISSUE THREAD

**Issue by raintonr** [raintonr/hormann-hcp#3]

**Comment by Bouni** [comment]
Just out of curiosity, why do you doubt that this will not work on a SBC?

**Comment by raintonr** [comment]
> Just out of curiosity, why do you doubt that this will not work on a SBC?

Timing issues? Never take it for granted is all.

**Comment by raintonr** [comment]
Oh... and also should get a TTL <-> RS485 converter and confirm it will work using the RPi's built in UART as think that will be my preferred method of use.

**Comment by Bouni** [comment]
Yesterday, late in the evening I found a RPi 4 to use for the test, but unfortunately it was to late (and to cold outside) to do an actual test. I Installed a fresh RPi OS, installed nodejs and npm, changed this line [hormann-hcp.js#L20](https://github.com/raintonr/hormann-hcp/blob/master/hormann-hcp.js#L20) to use `/dev/serial1` which should be the built in UART of the RPi.

So far everything worked, but I had no MAX3485 laying around to build a UART to RS485 adapter. 
Will do that in the next few days and hopefully do a live test.

**Comment by raintonr** [comment]
I have a TTL <-> 485 adapter on order (and a spare RPi for experiments).

I read one needs to be careful to choose adapter that runs at 3.3v so as not to fry RPi. Hopeful I got the right one!

**Comment by riogrande75** [comment]
Just gave it a shot. Unfortunately no respone on any of the commands.
Logfile attached - maybe you can explain from this whats wrong.
[SupraMaticE2log.txt](https://github.com/raintonr/hormann-hcp/files/5994924/SupraMaticE2log.txt)

**Comment by raintonr** [comment]
I imagine that's is a timing issue. What hardware is this using?

**Comment by riogrande75** [comment]
I worked quite a lot with rs485 stuff and was mostly very successful with this adapter:
https://de.aliexpress.com/item/4000834530016.html
Have same adapter as you too, will give it a try later.

**Comment by riogrande75** [comment]
No output at all with the "cheap" ch341 based adapter (same u're using).
Could you pls. add a debugging switch (e.g. to output rx messages)? Just to be sure that HW is ok.

**Comment by raintonr** [comment]
> Could you pls. add a debugging switch (e.g. to output rx messages)? Just to be sure that HW is ok.

You can see in the code there is a lot of debug, that is commented. Just remove the comment markers and you will see lots of output.

**Comment by riogrande75** [comment]
Ok, even I see signal on rs485 side of usb adapter with oscilloscope nothing is received in linux => cheap crap.
What can I do to move on with the other adapter? Should I give it a try with a linux laptop instead the orangepi?
Or is the timing issue more likely coming from the adapter?

**Comment by raintonr** [comment]
> What can I do to move on with the other adapter? Should I give it a try with a linux laptop instead the orangepi?
> Or is the timing issue more likely coming from the adapter?

Who knows. All I can tell you is it works fine with my X220 laptop and the adapter as discussed here: https://github.com/raintonr/hormann-hcp/issues/2#issuecomment-777278987

**Comment by riogrande75** [comment]
Ok, got it finally - good work Robin,thx!
RPi2 with actual Raspbian Buster and this adapter connected to hw rx/tx: https://de.aliexpress.com/item/32811446272.html
Serial port set to /dev/ttyAMA0 and script started.
By pressing 1 door opens (at least a bit), 2 closes the door, 5 turns on internal lightning. Not really flawless so far, I have to repeat it some times to get the door fully open/closed, but at least a start.

Still get tons of these messages:
```
+118867725            Slave status request     280120
                                Start packet with non-zero?! (27)
                                Start packet with non-zero?! (54)
                CRC: 0 != 44
                CRC: 0 != 235
                                Start packet with non-zero?! (255)
                                Start packet with non-zero?! (32)
                                Start packet with non-zero?! (251)
                                Start packet with non-zero?! (219)
                                Start packet with non-zero?! (251)
                                Start packet with non-zero?! (254)
                                Start packet with non-zero?! (128)
                                Start packet with non-zero?! (32)
                                Start packet with non-zero?! (27)
                                Start packet with non-zero?! (248)
                                Start packet with non-zero?! (252)
                                Start packet with non-zero?! (255)
                                Start packet with non-zero?! (255)
                                Start packet with non-zero?! (251)
                                Start packet with non-zero?! (255)
                CRC: 54 != 154

```
Do they all point to timing issues? Why does it work then? Maybe I should give it a try with a fast x64 pc to be sure, RPi2 is not to slow for this job.

**Comment by raintonr** [comment]
> Ok, got it finally - good work Robin,thx!

Cool :)

But...

> Still get tons of these messages:

Try messing with the `packetInterval` constant in the parser (https://github.com/raintonr/hormann-hcp/blob/master/parser-hcp.js#L5) to see if you can get rid of those messages.

For command sending... once sent, the command bits are then cleared, so if it's sent and the motor responds that should be good. I don't understand why the motor would move a bit and then stop, unless it's something like a gate and you have send the 'pedestrian open' command. Well... that was how it was with my Lineamatic P motor. Maybe these commands need holding high for other devices?

**Comment by riogrande75** [comment]
> Try messing with the `packetInterval` constant in the parser (https://github.com/raintonr/hormann-hcp/blob/master/parser-hcp.js#L5) to see if you can get rid of those messages.

Did that. So far no change noticeable, no matter if it use values between 2000000 and 2500000.
I tried a 120Ω resistor as a bus termination too, without success. Problem must be on RPi side not HW.

> For command sending... once sent, the command bits are then cleared, so if it's sent and the motor responds that should be good. I don't understand why the motor would move a bit and then stop, unless it's something like a gate and you have send the 'pedestrian open' command.

Guess that hast something to do with the problem as well.

> Well... that was how it was with my Lineamatic P motor. Maybe these commands need holding high for other devices?

No, don't think so. Since my Supramatic E2 works with UAP1 just like your Lineamatic P, I don't expect any difference in the protocol. Especially when having a look at the very limited ressources of the μP on the UAP1.

**Comment by raintonr** [comment]
> > Try messing with the `packetInterval` constant in the parser (https://github.com/raintonr/hormann-hcp/blob/master/parser-hcp.js#L5) to see if you can get rid of those messages.
> 
> Did that. So far no change noticeable, no matter if it use values between 2000000 and 2500000.
> I tried a 120Ω resistor as a bus termination too, without success. Problem must be on RPi side not HW.

I will make some changes so debug can be turned on/off with env. variable, and add intra-packet timeout logging. That way we will be able to see exact timings between send/received. It is conceivable that the RPi is not fast enough to decode and respond in time (response should be sent 3-10ms after incoming packet) but that doesn't really sound likely. I mean, 10ms is an eternity in computing terms.

**Comment by riogrande75** [comment]
>10ms is an eternity in computing terms.

Absolutely - I agree.
Really looking forward for these changes to get rid of the problems. Will help to diagnose issues on other platforms too.

I am curious on the findings of @Bouni with his RPi4.

**Comment by Bouni** [comment]
Hi, 
I still wait for my MAX3485 IC to build the interface.

@riogrande75 Do you have a oscilloscope or a logic analyzer that you can provide timing traces of a message?

I did some test yesterday with an ESP32 where I set the TX pin to be a GPIO, pull it LOW for 13bit lengths, set it HIGH again wait another bit length and then switch to Serial and send my bytes.

Looks promissing to use that technique to get a sync break.

Just forgot to record my logic traces and there fore would be nice to have actual traces from a actual communication to see the timings.

Edit:

In my opinion a RPi / OrangePi / SBC whatever is never ever to slow for a serial comunication with 19200 Baud, as that is awefully slow for a PC as @raintonr said earlier.

**Comment by riogrande75** [comment]
ESP based controller - yeah, that's it. I have more than just one door, so I could afford to control all of them.
Attached some traces with my logic analyzer (rx/tx at RPi / pulseview format) in parallel with the console output. Maybe @Bouni can see, why I got so many faults.

[Door.zip](https://github.com/raintonr/hormann-hcp/files/6015426/Door.zip)

**Comment by raintonr** [comment]
> RPi2 with actual Raspbian Buster and this adapter connected to hw rx/tx: https://de.aliexpress.com/item/32811446272.html

I have now got something similar, so I tested it with a regular USB/RS232 module I have. Ie. USB -> RS232 (wired to ->) TTL -> 485.

To begin got a whole load of errors of the type you saw, so I messed with the timing a bit. The `master` version of the code right now works with the setup I just described and the original USB -> 485 adapter I started with.

If you could have a go with the current code, that might solve your issues then?

> Serial port set to /dev/ttyAMA0 and script started.

~~I couldn't get any serial device to work on my RPi, device won't even open. Not the USB one (on `/dev/ttyUSB0`) or the built in one (on `/dev/ttyAMA0`). Get this error which I don't understand. Do you guys know what this means (don't worry about `root` - this is just a throw away test environment):~~ Scrub that - re-installed the node modules and the error below went away.

```
root@pp:~/src/hormann-hcp# node hormann-hcp.js 
/root/src/hormann-hcp/node_modules/bindings/bindings.js:121
        throw e;
        ^

Error: Error relocating /root/src/hormann-hcp/node_modules/@serialport/bindings/build/Release/bindings.node: unsupported relocation type 7
    at Object.Module._extensions..node (internal/modules/cjs/loader.js:1122:18)
    at Module.load (internal/modules/cjs/loader.js:928:32)
    at Function.Module._load (internal/modules/cjs/loader.js:769:14)
    at Module.require (internal/modules/cjs/loader.js:952:19)
    at require (internal/modules/cjs/helpers.js:88:18)
    at bindings (/root/src/hormann-hcp/node_modules/bindings/bindings.js:112:48)
    at Object.<anonymous> (/root/src/hormann-hcp/node_modules/@serialport/bindings/lib/linux.js:2:36)
    at Module._compile (internal/modules/cjs/loader.js:1063:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1092:10)
    at Module.load (internal/modules/cjs/loader.js:928:32)
```

**Comment by raintonr** [comment]
> Serial port set to /dev/ttyAMA0 and script started.
>...
> Still get tons of these messages:
> 
> ```
> +118867725            Slave status request     280120
>                                 Start packet with non-zero?! (27)
>                                 Start packet with non-zero?! (54)
>                 CRC: 0 != 44
>                 CRC: 0 != 235
>                                 Start packet with non-zero?! (255)
>                                 Start packet with non-zero?! (32)
>                                 Start packet with non-zero?! (251)
>                                 Start packet with non-zero?! (219)
>                                 Start packet with non-zero?! (251)
>                                 Start packet with non-zero?! (254)
>                                 Start packet with non-zero?! (128)
>                                 Start packet with non-zero?! (32)
>                                 Start packet with non-zero?! (27)
>                                 Start packet with non-zero?! (248)
>                                 Start packet with non-zero?! (252)
>                                 Start packet with non-zero?! (255)
>                                 Start packet with non-zero?! (255)
>                                 Start packet with non-zero?! (251)
>                                 Start packet with non-zero?! (255)
>                 CRC: 54 != 154
> ```

Now I have an RPi (3B+) running with TTL <-> 485 module connected to `/dev/ttyAMA0` I see the same thing.

The problem here is likely the break sequence. Looks like USB UARTs somehow interpret that as an error (just a guess) and spit out a zero byte but the RPi's built in UART doesn't do this. Looking at what's coming in... seems like that zero byte sometimes is thrown out and sometimes not. Will try and tweak the code accordingly.

**Comment by raintonr** [comment]
> The problem here is likely the break sequence. Looks like USB UARTs somehow interpret that as an error (just a guess) and spit out a zero byte but the RPi's built in UART doesn't do this. Looking at what's coming in... seems like that zero byte sometimes is thrown out and sometimes not. Will try and tweak the code accordingly.

All good now. Current code works on RPi through both internal UART and external USB device (tested with FTDI).

Still throws a few too many CRC errors for my liking but will work on that separately.

**Comment by raintonr** [comment]
> Still throws a few too many CRC errors for my liking but will work on that separately.

That is also fixed now.

@riogrande75 if you want to give this a test would be interested to see if it's working for you too.

**Comment by riogrande75** [comment]
Yes - much better, works pretty flawless now.
Great work -THX!
